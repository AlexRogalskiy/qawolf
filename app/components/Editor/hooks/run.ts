import { useRouter } from "next/router";
import { useEffect } from "react";

import { useEditor, useSuite } from "../../../hooks/queries";
import { state } from "../../../lib/state";
import { Run, Suite, Team } from "../../../lib/types";

export type RunHook = {
  run: Run | null;
  suite: Suite | null;
};

type UseRun = {
  branch?: string;
  runId?: string;
  team?: Team;
};

const pollInterval = 2000;

export const useRun = ({ branch, runId, team }: UseRun): RunHook => {
  const { query } = useRouter();

  // TODO replace with run query
  const { data, startPolling, stopPolling } = useEditor(
    { branch, run_id: runId },
    { teamId: team?.id }
  );
  const editorData = data?.editor || null;
  const run = editorData?.run || null;

  const { data: suiteData } = useSuite(
    { id: run?.suite_id || (query?.suite_id as string) },
    { teamId: team?.id }
  );
  const suite = suiteData?.suite || null;

  // tee up correct branch and environment if test edited
  useEffect(() => {
    if (suite?.environment_id) {
      state.setEnvironmentId(suite.environment_id);
    }

    if (suite?.branch && team.git_sync_integration_id) {
      state.setBranch(suite.branch);
    }
  }, [suite, team]);

  useEffect(() => {
    if (!run || run.completed_at) return;

    startPolling(pollInterval);

    return () => {
      stopPolling();
    };
  }, [run, startPolling, stopPolling]);

  return {
    run,
    suite,
  };
};
