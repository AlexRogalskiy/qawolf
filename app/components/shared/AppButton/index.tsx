import { Box, BoxProps, Button, ButtonProps } from "grommet";
import { Icon } from "grommet-icons";
import Link from "next/link";
import styled from "styled-components";

import { Side } from "../../../lib/types";
import {
  borderSize,
  colors,
  edgeSize,
  overflowStyle,
  transition,
  transitionDuration,
} from "../../../theme/theme";
import Text from "../Text";
import {
  activeBackground,
  activeSecondaryBackground,
  background,
  hoverBackground,
  hoverSecondaryBackground,
  textColor,
  Type,
} from "./config";
import { getBoxPad, getTextMargin } from "./helpers";

export type Props = {
  IconComponent?: Icon;
  a11yTitle?: string;
  className?: string;
  color?: string;
  hasError?: boolean;
  href?: string;
  hoverType?: Type;
  iconColor?: string;
  iconPosition?: Side;
  isDisabled?: boolean;
  isLarge?: boolean;
  isSelected?: boolean;
  justify?: BoxProps["justify"];
  label?: string;
  margin?: ButtonProps["margin"];
  noBorderSide?: Side;
  onClick?: () => void;
  openNewPage?: boolean;
  type: Type;
  width?: BoxProps["width"];
};

function AppButton({
  IconComponent,
  a11yTitle,
  className,
  color,
  href,
  iconColor,
  iconPosition,
  isDisabled,
  isLarge,
  justify,
  label,
  margin,
  onClick,
  openNewPage,
  type,
}: Props): JSX.Element {
  const innerHtml = (
    <Button
      a11yTitle={a11yTitle || label}
      className={className}
      disabled={isDisabled}
      margin={margin}
      onClick={onClick}
      plain
      style={{ transition }}
    >
      <Box
        align="center"
        direction={iconPosition === "right" ? "row-reverse" : "row"}
        fill
        justify={justify || "between"}
        pad={
          isLarge
            ? { horizontal: "small" }
            : getBoxPad({
                hasIcon: !!IconComponent,
                hasLabel: !!label,
                iconPosition,
                justify,
                type,
              })
        }
      >
        {!!IconComponent && (
          <IconComponent
            color={iconColor || color || textColor[type]}
            size={edgeSize.small}
          />
        )}
        {!!label && (
          <Text
            color={color || textColor[type]}
            margin={getTextMargin(!!IconComponent, iconPosition)}
            size={isLarge ? "buttonLarge" : "component"}
            style={overflowStyle}
          >
            {label}
          </Text>
        )}
      </Box>
    </Button>
  );

  if (href) {
    return (
      <Link href={href}>
        <a target={openNewPage ? "_blank" : undefined}>{innerHtml}</a>
      </Link>
    );
  }

  return innerHtml;
}

const StyledAppButton = styled(AppButton)`
  background: ${(props) => `${background[props.type]}`};
  border-radius: ${(props) =>
    props.isLarge ? borderSize.medium : borderSize.small};
  height: ${(props) => (props.isLarge ? edgeSize.xxlarge : edgeSize.large)};

  ${(props) => !!props.isDisabled && "cursor: not-allowed;"}
  ${(props) => !!props.width && `width: ${props.width};`}

  ${(props) =>
    props.type === "dark" &&
    `
  border: ${borderSize.xsmall} solid ${colors.gray8};
  `}

  ${(props) =>
    props.type === "disabled" &&
    `
  border: ${borderSize.xsmall} solid ${colors.gray3};
  `}

  ${(props) =>
    props.type === "secondary" &&
    `
  border: ${borderSize.xsmall} solid ${colors.gray3};
  `}

  ${(props) =>
    !!props.noBorderSide &&
    `
  border-${props.noBorderSide}: none;
  border-bottom-${props.noBorderSide}-radius: 0;
  border-top-${props.noBorderSide}-radius: 0;
  `}

  ${(props) => !!props.hasError && `border-color: ${colors.danger5};`}
  ${(props) => !!props.isSelected && `border-color: ${colors.gray9};`}

  svg {
    transition: fill ${transitionDuration};
  }

  &:hover {
    ${(props) =>
      `
    background: ${
      hoverSecondaryBackground[props.hoverType] || hoverBackground[props.type]
    };

    p, 
    svg {
      color: ${textColor[props.hoverType || props.type]};
      fill: ${textColor[props.hoverType || props.type]};
    }
    `}

    ${(props) => props.type === "dark" && `border-color: ${colors.gray6};`}
    ${(props) => props.type === "secondary" && `border-color: ${colors.gray5};`}
  }

  &:active {
    ${(props) => `
    background: ${
      activeSecondaryBackground[props.hoverType] || activeBackground[props.type]
    };
    `}

    ${(props) => props.type === "dark" && `border-color: ${colors.gray4};`}
    ${(props) => props.type === "secondary" && `border-color: ${colors.gray9};`}
  }
`;

export default StyledAppButton;
