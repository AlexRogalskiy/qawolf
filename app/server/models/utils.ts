export const COLORS = [
  "#4545E5",
  "#C54BDE",
  "#56BBD6",
  "#8BC22D",
  "#E59C59",
  "#DA4E94",
  "#ABB3C2",
  "#667080",
];

export const buildColor = (
  teamColors: string[],
  colors: string[] = COLORS
): string => {
  const availableColor = colors.find((color) => !teamColors.includes(color));

  if (availableColor) return availableColor;

  return colors[teamColors.length % colors.length];
};
