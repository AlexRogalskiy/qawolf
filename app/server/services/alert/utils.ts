import { Suite, Trigger } from "../../types";

type BuildSuiteName = {
  suite: Suite;
  trigger: Trigger | null;
};

export const buildSuiteName = ({ suite, trigger }: BuildSuiteName): string => {
  return trigger?.name || suite.tag_names || "manually triggered";
};
