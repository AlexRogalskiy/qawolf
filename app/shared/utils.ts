export const buildTestCode = (
  url: string,
  domcontentloaded = false
): string => {
  let code =
    "const { context } = await launch();\nconst page = await context.newPage();\n";

  code += domcontentloaded
    ? `await page.goto('${url}', { waitUntil: "domcontentloaded" })\n`
    : `await page.goto('${url}');\n`;

  code += "// 🐺 QA Wolf will create code here";

  return code;
};

export const daysFromNow = (days = 0, existingDate?: number): string => {
  const date = existingDate ? new Date(existingDate) : new Date();
  date.setDate(date.getDate() + days);

  return date.toISOString();
};

// https://stackoverflow.com/a/1214753
export const minutesFromNow = (minutes = 0): string => {
  return new Date(Date.now() + minutes * 60000).toISOString();
};
