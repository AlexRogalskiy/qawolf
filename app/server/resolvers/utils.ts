/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { Transaction } from "knex";

import { AuthenticationError } from "../errors";
import { Logger } from "../Logger";
import { findEnvironment } from "../models/environment";
import { findEnvironmentVariable } from "../models/environment_variable";
import { findSuite } from "../models/suite";
import { findTest } from "../models/test";
import { findTrigger } from "../models/trigger";
import { Team, Test, User } from "../types";

type EnsureEnvironmentAccess = {
  environment_id: string;
  logger: Logger;
  teams: Team[] | null;
  trx?: Transaction;
};

type EnsureEnvironmentVariableAccess = {
  environment_variable_id: string;
  logger: Logger;
  teams: Team[] | null;
  trx?: Transaction;
};

type EnsureSuiteAccess = {
  logger: Logger;
  suite_id: string;
  teams: Team[] | null;
};

type EnsureTeams = {
  logger: Logger;
  teams: Team[] | null;
};

type EnsureTeamAccess = {
  logger: Logger;
  team_id: string;
  teams: Team[] | null;
};

type EnsureTestAccess = {
  logger: Logger;
  teams: Team[] | null;
  test?: Test;
  test_id?: string;
};

type EnsureTriggerAccess = {
  logger: Logger;
  teams: Team[] | null;
  trigger_id: string;
  trx?: Transaction;
};

type EnsureUser = {
  logger: Logger;
  user: User | null;
};

/**
 * @summary Throws if `teams` is falsy
 * @returns `teams`
 */
export const ensureTeams = ({ logger, teams }: EnsureTeams): Team[] => {
  const log = logger.prefix("ensureTeams");

  if (!teams) {
    log.error("teams not provided");
    throw new AuthenticationError("no teams");
  }

  return teams;
};

export const ensureEnvironmentAccess = async ({
  environment_id,
  logger,
  teams,
  trx,
}: EnsureEnvironmentAccess): Promise<Team> => {
  const log = logger.prefix("ensureEnvironmentAccess");

  const teamIds = ensureTeams({ teams, logger }).map((team) => team.id);
  log.debug("ensure teams", teamIds, "can access environment", environment_id);

  const environment = await findEnvironment(environment_id, { logger, trx });
  const teamForEnvironment = teams!.find(
    (team) => environment.team_id === team.id
  );

  if (!teamForEnvironment) {
    log.error("teams", teamIds, "cannot access environment", environment_id);
    throw new AuthenticationError("cannot access environment");
  }

  return teamForEnvironment;
};

export const ensureEnvironmentVariableAccess = async ({
  environment_variable_id,
  logger,
  teams,
  trx,
}: EnsureEnvironmentVariableAccess): Promise<Team> => {
  const log = logger.prefix("ensureEnvironmentVariableAccess");

  const teamIds = ensureTeams({ teams, logger }).map((team) => team.id);
  log.debug(
    "ensure teams",
    teamIds,
    "can access environment variable",
    environment_variable_id
  );

  const variable = await findEnvironmentVariable(environment_variable_id, {
    logger,
    trx,
  });
  const selectedTeam = teams!.find((team) => variable.team_id === team.id);

  if (!selectedTeam) {
    log.error(
      "teams",
      teamIds,
      "cannot access environment variable",
      environment_variable_id
    );
    throw new AuthenticationError("cannot access environment variable");
  }

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `suite.team_id`. Can also
 *   throw if `suite_id` doesn't match any suite.
 * @returns The team from `teams` that matches `suite.team_id`, if we didn't throw.
 */
export const ensureSuiteAccess = async ({
  logger,
  teams,
  suite_id,
}: EnsureSuiteAccess): Promise<void> => {
  const log = logger.prefix("ensureSuiteAccess");

  const teamIds = ensureTeams({ logger, teams }).map((team) => team.id);

  log.debug("ensure teams", teamIds, "can access suite", suite_id);

  const suite = await findSuite(suite_id, { logger });

  if (!teamIds.includes(suite.team_id)) {
    log.error("teams", teamIds, "cannot access suite", suite_id);
    throw new Error("Teams cannot access suite");
  }
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `team_id`
 * @returns The team from `teams` that matches `team_id`, if we didn't throw.
 */
export const ensureTeamAccess = ({
  logger,
  team_id,
  teams,
}: EnsureTeamAccess): Team => {
  const log = logger.prefix("ensureTeamAccess");

  const teamIds = ensureTeams({ logger, teams }).map((team) => team.id);

  const selectedTeam = teams!.find((team) => team.id === team_id);
  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access team", team_id);
    throw new AuthenticationError("cannot access team");
  }

  log.debug("ensured teams", teamIds, "can access team", team_id);

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `test.team_id`. Can also
 *   throw if `test_id` doesn't match any test.
 * @returns The team from `teams` that matches `test.team_id`, if we didn't throw.
 */
export const ensureTestAccess = async ({
  logger,
  teams,
  test,
  test_id,
}: EnsureTestAccess): Promise<Team> => {
  const log = logger.prefix("ensureTestAccess");

  const finalTeams = ensureTeams({ logger, teams });
  const teamIds = finalTeams.map((t) => t.id);

  log.debug("ensure teams", teamIds, "can access test", test?.id || test_id);

  const finalTest = test || (await findTest(test_id, { logger }));
  const team = finalTeams.find((t) => finalTest.team_id === t.id);

  if (!team) {
    log.error("teams", teamIds, "cannot access test", test?.id || test_id);
    throw new AuthenticationError("cannot access test");
  }

  return team;
};

export const ensureTriggerAccess = async ({
  logger,
  teams,
  trigger_id,
  trx,
}: EnsureTriggerAccess): Promise<Team> => {
  const log = logger.prefix("ensureTriggerAccess");

  const teamIds = ensureTeams({ teams, logger }).map((team) => team.id);
  log.debug("ensure teams", teamIds, "can access trigger", trigger_id);

  const trigger = await findTrigger(trigger_id, { logger, trx });
  const selectedTeam = teams!.find((team) => trigger.team_id === team.id);

  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access trigger", trigger_id);
    throw new AuthenticationError("cannot access trigger");
  }

  return selectedTeam;
};

/**
 * @summary Throws if `user` is falsy
 * @returns `user` if we didn't throw
 */
export const ensureUser = ({ logger, user }: EnsureUser): User => {
  const log = logger.prefix("ensureUser");

  if (!user) {
    log.error("user not provided");
    throw new AuthenticationError("no user");
  }

  return user;
};
