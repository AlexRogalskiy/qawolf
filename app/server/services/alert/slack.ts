import { IncomingWebhook, IncomingWebhookSendArguments } from "@slack/webhook";
import last from "lodash/last";

import environment from "../../environment";
import { findIntegration } from "../../models/integration";
import { findUsersForTeam } from "../../models/user";
import { ModelOptions, Suite, SuiteRun, Trigger, User } from "../../types";
import { randomChoice } from "../../utils";

type BuildSuiteMessage = {
  runs: SuiteRun[];
  suite: Suite;
  trigger: Trigger | null;
  user: User | null;
};

type PostMessageToSlack = {
  message: IncomingWebhookSendArguments;
  webhook_url: string;
};

type SendSlackAlert = {
  integrationId: string;
  runs: SuiteRun[];
  suite: Suite;
  trigger: Trigger | null;
};

const buildSuiteDetailBlocks = (
  suite: Suite,
  isPass: boolean
): IncomingWebhookSendArguments["blocks"] => {
  if (!suite.commit_url) return [];

  const emoji = isPass ? "✅" : "❌";
  const sha = last(suite.commit_url.split("/")).slice(0, 7);
  const pullRequestId = suite.pull_request_url
    ? last(suite.pull_request_url.split("/"))
    : null;

  const branch = suite.branch ? `${suite.branch}, ` : "";
  const pullRequest = pullRequestId
    ? ` (<${suite.pull_request_url}|#${pullRequestId}>)`
    : "";

  return [
    {
      type: "section",
      text: {
        type: "mrkdwn",
        text: `${emoji} <${suite.commit_url}|${sha}>: ${branch}${
          suite.commit_message || ""
        }${pullRequest}`,
      },
    },
  ];
};

export const buildMessageForSuite = ({
  runs,
  suite,
  trigger,
  user,
}: BuildSuiteMessage): IncomingWebhookSendArguments => {
  const wolfName = user?.wolf_name || "Spirit";
  const wolfVariant = user?.wolf_variant || "white";

  const failingRuns = runs.filter((r) => r.status === "fail");
  const status = failingRuns.length ? "failed." : "passed!";

  const triggerName = trigger?.name || "manually triggered";

  const suiteHref = new URL(`/suites/${suite.id}`, environment.APP_URL).href;
  const headline = `${wolfName} here: <${suiteHref}|${triggerName} tests> ${status}`;
  const text = `${triggerName} tests ${status}`;

  const runBlocks = failingRuns.map((run) => {
    return {
      type: "section",
      text: {
        text: `<${environment.APP_URL}/run/${run.id}|${run.test_name}>`,
        type: "mrkdwn",
      },
      accessory: {
        alt_text: run.test_name,
        image_url: run.gif_url,
        type: "image",
      },
    };
  });

  return {
    blocks: [
      {
        type: "context",
        elements: [
          {
            type: "image",
            image_url: `https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-${wolfVariant}${
              failingRuns.length ? "" : "-party"
            }-slack.png`,
            alt_text: wolfName,
          },
          {
            type: "mrkdwn",
            text: headline,
          },
        ],
      },
      ...buildSuiteDetailBlocks(suite, !failingRuns.length),
      ...runBlocks,
    ],
    text,
  };
};

export const postMessageToSlack = async ({
  message,
  webhook_url,
}: PostMessageToSlack): Promise<void> => {
  const webhook = new IncomingWebhook(webhook_url);
  await webhook.send(message);
};

export const sendSlackAlert = async (
  { integrationId, runs, suite, trigger }: SendSlackAlert,
  options: ModelOptions
): Promise<void> => {
  const log = options.logger.prefix("sendSlackAlert");
  log.debug("suite", suite.id);

  try {
    const users = await findUsersForTeam(suite.team_id, options);
    const integration = await findIntegration(integrationId, options);

    if (!integration.webhook_url) {
      log.error("no webhook url for integration", integration.id);
      throw new Error("No webhook url for Slack");
    }

    const user = users.length ? (randomChoice(users) as User) : null;

    await postMessageToSlack({
      message: buildMessageForSuite({ runs, suite, trigger, user }),
      webhook_url: integration.webhook_url,
    });
    log.debug("sent");
  } catch (error) {
    log.alert("error: Slack alert", error.message);
  }
};
