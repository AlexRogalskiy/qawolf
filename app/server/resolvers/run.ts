import { db } from "../db";
import environment from "../environment";
import { AuthenticationError } from "../errors";
import { Logger } from "../Logger";
import { findRun, findRunsForSuite, UpdateRun, updateRun } from "../models/run";
import { expireRunner, findRunner } from "../models/runner";
import { postMessageToSlack } from "../services/alert/slack";
import {
  Context,
  ModelOptions,
  Run,
  Suite,
  SuiteRun,
  UpdateRunMutation,
} from "../types";
import { minutesFromNow } from "../utils";
import { ensureSuiteAccess } from "./utils";

type ValidateApiKey = {
  api_key: string | null;
  run: Run;
};

const postRunFailedMessageToSlack = async (
  run: Run,
  logger: Logger
): Promise<void> => {
  if (!environment.SLACK_UPDATES_WEBHOOK) return;

  const log = logger.prefix("postRunFailedMessageToSlack");

  try {
    await postMessageToSlack({
      message: {
        text: `ðŸš¨ Run failure: run ${run.id} failed (test ${run.test_id})!`,
      },
      webhook_url: environment.SLACK_UPDATES_WEBHOOK,
    });

    log.debug("sent");
  } catch (error) {
    log.alert("could not send slack message", error.message);
  }
};

/**
 * @summary Checks that an API key is valid and allows mutating
 *   the given run.
 */
export const validateApiKey = async (
  { api_key, run }: ValidateApiKey,
  { logger, trx }: ModelOptions
): Promise<void> => {
  const log = logger.prefix("validateApiKey");
  log.debug();

  if (!api_key) {
    log.error("no api key provided");
    throw new AuthenticationError("invalid api key");
  }

  const runner = await findRunner({ run_id: run.id }, { logger, trx });

  if (!runner || api_key !== runner.api_key) {
    log.error("incorrect api key for run", run?.id, "runner", runner?.id);
    throw new AuthenticationError("invalid api key");
  }
};

/**
 * @returns Array of SuiteRuns
 */
export const suiteRunsResolver = async (
  { id }: Suite,
  _: Record<string, unknown>,
  { logger, teams }: Context
): Promise<SuiteRun[]> => {
  logger.debug("suiteRunsResolver");
  await ensureSuiteAccess({ logger, teams, suite_id: id });

  return findRunsForSuite(id, { logger });
};

/**
 * @returns Updated run record
 */
export const updateRunResolver = async (
  _: Record<string, unknown>,
  { current_line, id, status }: UpdateRunMutation,
  { api_key, logger }: Context
): Promise<Run> => {
  const log = logger.prefix("updateRunResolver");
  log.debug(id);

  let postMessagePromise: Promise<unknown> = Promise.resolve();

  const updatedRun = await db.transaction(async (trx) => {
    const run = await findRun(id, { logger, trx });

    await validateApiKey({ api_key, run }, { logger, trx });

    if (status === "fail") {
      // wait up to 100ms for message to post
      postMessagePromise = Promise.race([
        postRunFailedMessageToSlack(run, logger),
        new Promise((resolve) => setTimeout(resolve, 100)),
      ]);
    }

    const updates: UpdateRun = { id };

    if (status === "created") {
      updates.started_at = minutesFromNow();
    } else {
      updates.current_line = current_line;
      updates.status = status;
    }

    if (status === "pass") {
      await expireRunner({ run_id: id }, { logger, trx });
    }

    return updateRun(updates, { logger });
  });

  await postMessagePromise;

  return updatedRun;
};
