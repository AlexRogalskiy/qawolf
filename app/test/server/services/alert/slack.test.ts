import { db, dropTestDb, migrateDb } from "../../../../server/db";
import { findRunsForSuite } from "../../../../server/models/run";
import * as slack from "../../../../server/services/alert/slack";
import * as azure from "../../../../server/services/aws/storage";
import { minutesFromNow } from "../../../../shared/utils";
import {
  buildIntegration,
  buildRun,
  buildSuite,
  buildTeam,
  buildTeamUser,
  buildTest,
  buildTrigger,
  buildUser,
  logger,
} from "../../utils";

const { buildMessageForSuite, sendSlackAlert } = slack;

const trigger = buildTrigger({});
const integration = buildIntegration({});
const user = buildUser({});

beforeAll(async () => {
  await migrateDb();

  await db("users").insert(user);
  await db("teams").insert(buildTeam({}));
  await db("team_users").insert(buildTeamUser({}));

  await db("integrations").insert(integration);
  await db("teams").update({ alert_integration_id: "integrationId" });

  await db("triggers").insert(trigger);

  await db("suites").insert([buildSuite({})]);
  await db("tests").insert(buildTest({ name: "testName" }));
  return db("runs").insert([
    buildRun({
      completed_at: minutesFromNow(),
      status: "fail",
      suite_id: "suiteId",
    }),
    buildRun({
      completed_at: minutesFromNow(),
      i: 2,
      status: "pass",
      suite_id: "suiteId",
    }),
  ]);
});

afterAll(() => dropTestDb());

describe("buildMessageForSuite", () => {
  beforeAll(() => {
    jest.spyOn(azure, "createStorageReadAccessUrl").mockReturnValue("url");
  });

  afterAll(jest.restoreAllMocks);

  it("builds a Slack message for a failing suite", async () => {
    const trigger = await db.select("*").from("triggers").first();
    const suite = await db.select("*").from("suites").first();
    const runs = await findRunsForSuite(suite.id, { logger });

    expect(
      buildMessageForSuite({ runs, suite, trigger, user })
    ).toMatchSnapshot();
  });

  it("builds a Slack message for a passing suite", async () => {
    await db("runs").where({ id: "runId" }).update({ status: "pass" });

    const trigger = await db.select("*").from("triggers").first();
    const suite = await db.select("*").from("suites").first();
    const runs = await findRunsForSuite(suite.id, { logger });

    expect(
      buildMessageForSuite({ runs, suite, trigger, user })
    ).toMatchSnapshot();

    await db("runs").where({ id: "runId" }).update({ status: "fail" });
  });
});

describe("sendSlackAlert", () => {
  beforeAll(() => {
    jest.spyOn(slack, "postMessageToSlack").mockResolvedValue();
  });

  afterEach(jest.clearAllMocks);

  it("sends Slack alert", async () => {
    const trigger = await db.select("*").from("triggers").first();
    const suite = await db.select("*").from("suites").first();
    const runs = await findRunsForSuite(suite.id, { logger });

    await sendSlackAlert({
      integrationId: "integrationId",
      logger,
      runs,
      suite,
      trigger,
    });

    expect(slack.postMessageToSlack).toBeCalledWith({
      message: buildMessageForSuite({ runs, suite, trigger, user }),
      webhook_url: integration.webhook_url,
    });
  });
});
