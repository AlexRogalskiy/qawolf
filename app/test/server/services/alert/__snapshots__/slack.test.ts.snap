// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildMessageForSuite builds a Slack message for a failing suite 1`] = `
Object {
  "blocks": Array [
    Object {
      "elements": Array [
        Object {
          "alt_text": "Wolf1",
          "image_url": "https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-brown-slack.png",
          "type": "image",
        },
        Object {
          "text": "Wolf1 here: <http://localhost:3000/suites/suiteId|trigger1 tests> failed.",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
    Object {
      "accessory": Object {
        "alt_text": "testName",
        "image_url": "url",
        "type": "image",
      },
      "text": Object {
        "text": "<http://localhost:3000/run/runId|testName>",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
  "text": "trigger1 tests failed.",
}
`;

exports[`buildMessageForSuite builds a Slack message for a failing suite with git details 1`] = `
Object {
  "blocks": Array [
    Object {
      "elements": Array [
        Object {
          "alt_text": "Wolf1",
          "image_url": "https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-brown-slack.png",
          "type": "image",
        },
        Object {
          "text": "Wolf1 here: <http://localhost:3000/suites/suiteId|trigger1 tests> failed.",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
    Object {
      "text": Object {
        "text": "❌ <https://github.com/qawolf/example/pull/123/commits/af2565c5cc30fac7bccfcf442994f6324da4fcd9|af2565c>: feature, initial commit (<https://github.com/qawolf/example/pull/123|#123>)",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    Object {
      "accessory": Object {
        "alt_text": "testName",
        "image_url": "url",
        "type": "image",
      },
      "text": Object {
        "text": "<http://localhost:3000/run/runId|testName>",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
  "text": "trigger1 tests failed.",
}
`;

exports[`buildMessageForSuite builds a Slack message for a failing suite with git details 2`] = `
Object {
  "blocks": Array [
    Object {
      "elements": Array [
        Object {
          "alt_text": "Wolf1",
          "image_url": "https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-brown-slack.png",
          "type": "image",
        },
        Object {
          "text": "Wolf1 here: <http://localhost:3000/suites/suiteId|trigger1 tests> failed.",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
    Object {
      "text": Object {
        "text": "❌ <https://github.com/qawolf/example/pull/123/commits/af2565c5cc30fac7bccfcf442994f6324da4fcd9|af2565c>: feature, initial commit",
        "type": "mrkdwn",
      },
      "type": "section",
    },
    Object {
      "accessory": Object {
        "alt_text": "testName",
        "image_url": "url",
        "type": "image",
      },
      "text": Object {
        "text": "<http://localhost:3000/run/runId|testName>",
        "type": "mrkdwn",
      },
      "type": "section",
    },
  ],
  "text": "trigger1 tests failed.",
}
`;

exports[`buildMessageForSuite builds a Slack message for a passing suite 1`] = `
Object {
  "blocks": Array [
    Object {
      "elements": Array [
        Object {
          "alt_text": "Wolf1",
          "image_url": "https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-brown-party-slack.png",
          "type": "image",
        },
        Object {
          "text": "Wolf1 here: <http://localhost:3000/suites/suiteId|trigger1 tests> passed!",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
  ],
  "text": "trigger1 tests passed!",
}
`;

exports[`buildMessageForSuite builds a Slack message for a passing suite 2`] = `
Object {
  "blocks": Array [
    Object {
      "elements": Array [
        Object {
          "alt_text": "Wolf1",
          "image_url": "https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-brown-party-slack.png",
          "type": "image",
        },
        Object {
          "text": "Wolf1 here: <http://localhost:3000/suites/suiteId|Sign up tests> passed!",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
  ],
  "text": "Sign up tests passed!",
}
`;

exports[`buildMessageForSuite builds a Slack message for a passing suite 3`] = `
Object {
  "blocks": Array [
    Object {
      "elements": Array [
        Object {
          "alt_text": "Wolf1",
          "image_url": "https://qawolf-public.s3.us-east-2.amazonaws.com/wolf-brown-party-slack.png",
          "type": "image",
        },
        Object {
          "text": "Wolf1 here: <http://localhost:3000/suites/suiteId|manually triggered tests> passed!",
          "type": "mrkdwn",
        },
      ],
      "type": "context",
    },
  ],
  "text": "manually triggered tests passed!",
}
`;
